### PR TITLE
Datetime index: Introduce `RangeInterface`

### PR DESCRIPTION
**Pull request by coszio**
Initially it made sense to me to keep different types separate in `FieldCondition`, but now I propose to handle all `Range` conditions similarly and together, so here we add a new `RangeInterface` enum to wrap them together.

This change also made me realize we missed using the actual index sometimes when filtering `datetime_range` conditions (this is fixed now, but not very pretty, see `PayloadFieldIndex` implementation).

Note: this only affects the REST interface, GRPC is intact.

A bonus change I did is to use pattern matching and destructuring in the checkers section, so it will be checked by compiler to not forget any new conditions.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/index/field_index/numeric_index/mod.rs [hunk 1]
```diff
@@ -196,26 +197,50 @@ impl<T: Encodable + Numericable + Default> NumericIndex<T> {
...
+
+                let gbound = if let Some(gte) = datetime_range.gte {
+                    Included(T::from_i64(gte.timestamp_micros()))
+                } else if let Some(gt) = datetime_range.gt {
+                    Excluded(T::from_i64(gt.timestamp_micros()))
+                } else {
+                    Unbounded
+                };
+                (lbound, gbound)
+            }
```
Comment (xzfc) [inline lib/segment/src/index/field_index/numeric_index/mod.rs:243]: You could reduce code duplication by converting to `Range<T>` first.
Also, an idea: add `to_bounds` method to the `Range<T>` structure.

```rust
impl<T: Copy> Range<T> {
    fn to_bounds(&self) -> (Bound<T>, Bound<T>) {...}
}
```

So this code could be reduced to this:
```rust
let (lbound, gbound) = match range {
    RangeInterface::Float(range) => range.map(T::from_f64),
    RangeInterface::DateTime(range) => range.map(|t| T::from_i64(i.timestamp_micros())),
}.to_bounds();
```

File: lib/segment/src/index/field_index/numeric_index/mod.rs [hunk 2]
```diff
@@ -301,30 +326,62 @@ impl<T: Encodable + Numericable + Default> PayloadFieldIndex for NumericIndex<T>
...
+                        let v: T = T::from_i64(lt.timestamp_micros());
+                        Excluded(NumericIndexKey::new(v, PointOffsetType::MIN))
+                    }
+                    Range { lte: Some(lte), .. } => {
+                        let v: T = T::from_i64(lte.timestamp_micros());
+                        Included(NumericIndexKey::new(v, PointOffsetType::MAX))
+                    }
+                    _ => Unbounded,
+                };
+                (start_bound, end_bound)
```
Comment (xzfc) [inline lib/segment/src/index/field_index/numeric_index/mod.rs:383]: You could reduce code duplication by converting to `Range<T>` first.
```rust
let cond_range = match cond_range {
    RangeInterface::Float(range) => range.map(T::from_f64),
    RangeInterface::DateTime(range) => range.map(|t| T::from_i64(i.timestamp_micros())),
}
let start_bound = match cond_range {
...
```

**Review by xzfc — DISMISSED** [review]

File: lib/segment/src/index/field_index/numeric_index/mod.rs (hunk 1 above)
Comment (coszio) [reply lib/segment/src/index/field_index/numeric_index/mod.rs:243]: I actually implemented a similar thing for #3493, I'm going to reuse it here 👍

File: lib/segment/src/types.rs [hunk 3]
```diff
@@ -1323,6 +1323,29 @@ impl From<Vec<IntPayloadType>> for MatchExcept {
...
+            _ => None,
+        }
+    }
+
+    pub fn to_datetime_range(self) -> Option<Range<DateTimePayloadType>> {
+        match self {
+            RangeInterface::DateTime(range) => Some(range),
+            _ => None,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/segment/src/types.rs:1346]: Good eye 👁️, I was using these for conversions, but then reimplemented them and forgot to remove them

File: lib/segment/src/index/query_optimization/condition_converter.rs [hunk 4]
```diff
@@ -128,55 +128,41 @@ pub fn field_condition_index<'a>(
...
+            ..
+        } => get_geo_polygon_checkers(index, geo_polygon.clone()),
 
-    if let Some(checker) = field_condition
-        .geo_polygon
-        .clone()
-        .and_then(|cond| get_geo_polygon_checkers(index, cond))
-    {
-        return Some(checker);
+        FieldCondition {
```
Comment (generall) [inline lib/segment/src/index/query_optimization/condition_converter.rs:156]: this is nice!

**Review by generall — APPROVED** [review]

**Pull request merged**